subprojects/docs/src/samples/userguide/dependencyManagement/managingTransitiveDependencies/declaringCapabilities/kotlin/build.gradle.kts

```diff
@@ -84,10 +84,7 @@ if (project.hasProperty("replace")) {
     // tag::use_slf4j[]
     configurations.all {
         resolutionStrategy.capabilitiesResolution.withCapability("log4j:log4j") {
-            select(candidates.find {
-                it as ModuleComponentIdentifier
-                it.module == "log4j-over-slf4j"
-            } )
+            select(candidates.find { it.getId<ModuleComponentIdentifier>().module == "log4j-over-slf4j" } )
```

Comment:
This code, and the previous one now that I think more about it, look very suspicious. But it could be my limited Kotlin knowledge.
What happens if one of the candidates is a `ProjectComponentIdentifier`? Does it ignore it or does it fail with a class cast exception?

Reply:
Yes it will give you a class cast exception. But the Groovy equivalent of the example will do the same (field not found exception) which is looking for the `id` field dynamically.
So I thought the new version of the example is the better alternative as the one before.
Because it at least indicates to users that there are subtypes.
Otherwise I think you are quite lost as a users.

I also think that the `ModuleComponentIdentifier` version is the one that you will need most of the time (we also have no test coverage for other cases). So my thinking was to make it a bit more elegant for that use cases, for which we actually have real world use cases.

But we could probably add more type safe APIs in addition. Although I would like to keep it "simple" for now.

I am fine to revert the example (and remove the type parameter) for now. Maybe it is misleading.

Reply:
I removed the type parameter and instead updated the samples to be type safe.